### PR TITLE
Add mercury layer

### PR DESCRIPTION
**Pull request by lboklin**
I created this a while ago while exploring the language and I figure there might be others looking for some manner of tooling for the language. 

It uses metal-mercury-mode from github rather than the mode available in MELPA because that one frankly doesn't work very well.

### CODE REVIEW

File: layers/+lang/mercury-layer/README.org [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#+TITLE: mercury layer
+
+# The maximum height of the logo should be 200 pixels.
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/README.org]: Please remove this line its just for introduction into the layer and should not be left in the finished layer.

File: layers/+lang/mercury-layer/README.org [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+[[file:img/mercury.png]]
+
+# TOC links should be GitHub style anchors.
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/README.org]: Same here

File: layers/+lang/mercury-layer/README.org [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+- [[#install][Install]]
+- [[#key-bindings][Key bindings]]
+
+* Description
+This layer adds support for the Mercury language.
+
+** Features:
+  - Indentation.
+  - Syntax highlighting.
+  - Compiling / compiling and running.
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/README.org]: Please change to:
```suggestion
  - Compiling and running.
```

File: layers/+lang/mercury-layer/README.org [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+file.
+
+* Key bindings
+
+| Key Binding | Description                          |
+|-------------+--------------------------------------|
+| ~SPC c c~   | Compile current buffer file.         |
+| ~SPC c r~   | Compile and run current buffer file. |
+
+# Use GitHub URLs if you wish to link a Spacemacs documentation file or its heading.
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/README.org]: Same here, please remove the comment lines

File: layers/+lang/mercury-layer/packages.el [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+;; Copyright (c) 2012-2018 Sylvain Benner & Contributors
+;;
+;; Author: Ludvig Böklin <ludvig.boklin@protonmail.com>
+;; URL: https://github.com/syl20bnr/spacemacs
+;;
+;; This file is not part of GNU Emacs.
+;;
+;;; License: GPLv3
+
+;;; Commentary:
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/packages.el]: This is just commentary for the layer author and should be removed.

File: layers/+lang/mercury-layer/packages.el [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+;; added to `mercury-packages'. Then, for each package PACKAGE:
+;;
+;; - If PACKAGE is not referenced by any other Spacemacs layer, define a
+;;   function `mercury/init-PACKAGE' to load and initialize the package.
+
+;; - Otherwise, PACKAGE is already referenced by another Spacemacs layer, so
+;;   define the functions `mercury/pre-init-PACKAGE' and/or
+;;   `mercury/post-init-PACKAGE' to customize the package as it is loaded.
+
+;;; Code:
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/packages.el]: From here the your code begins please remove all comments between line 39 and line 12

File: layers/+lang/mercury-layer/packages.el [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+(defconst mercury-packages
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/packages.el]: This should not be a defconst but a setq:
```suggestion
(setq mercury-packages ...
```

File: layers/+lang/mercury-layer/packages.el [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    smartparens))
+
+(defun mercury/post-init-flycheck ()
+  (add-hook 'metal-mercury-mode-hook 'flycheck-mode))
+
+(defun mercury/init-flycheck-mercury ()
+  "Initialize flycheck-mercury"
+  (use-package flycheck-mercury
+    :defer t
+    :init (require 'flycheck-mercury)))
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/packages.el]: Never use require in an init block, this renders your defer use-less.
Rather use a hook to load flycheck with mercury mode.

File: layers/+lang/mercury-layer/packages.el [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        ;; make
+        "cb" 'metal-mercury-mode-compile
+        "cr" 'metal-mercury-mode-runner))))
+
+(defun mercury/post-init-smartparens ()
+  (if dotspacemacs-smartparens-strict-mode
+      (add-hook 'metal-mercury-mode-hook #'smartparens-strict-mode)
+    (add-hook 'metal-mercury-mode-hook #'smartparens-mode)))
+
+;;; packages.el ends here
```
Comment (smile13241324) [inline layers/+lang/mercury-layer/packages.el]: No end package declaration for a layer file please.

File: layers/+lang/mercury-layer/packages.el (hunk 8 above)
Comment (lboklin) [reply layers/+lang/mercury-layer/packages.el]: Thanks for the feedback. I'm not very proficient in elisp, would you mind specifying in more detail where I should add the hook? In `mercury/init-metal-mercury-mode`?

**Comment by lboklin** [comment]
@smile13241324 I committed changes based on your feedback which resolves all but one of your points. Should I squash my commits once all changes have been made?

File: layers/+lang/mercury-layer/packages.el (hunk 8 above)
Comment (smile13241324) [reply layers/+lang/mercury-layer/packages.el]: Sure, you should add it to the init code of flycheck-mercury something among the line of
```elisp
(defun mercury/init-flycheck-mercury ()
  "Initialize flycheck-mercury"
  (use-package flycheck-mercury
    :defer t
    :init (add-hook 'metal-mercury-mode-hook Your-init-function)))
```
The idea is not to require the package but rather call one of the functions to autoload this package. Which function this is depends on the package. 

After checking the package in question I have seen that there is no autoload defined in there, so for now I would use the following code to init the package:
```elisp
(defun mercury/init-flycheck-mercury ()
  "Initialize flycheck-mercury"
  (use-package flycheck-mercury))
```

This should load the package immediately, we can later tweak it by adding autoloads from the layer.

**Comment by smile13241324** [comment]
@lboklin just checked your PR, looks good so far, I have added a more detailed comment to explain what to do with loading `flycheck-mercury`. When you are done please squash your commits and mention me again.

**Comment by smile13241324** [comment]
Thank you for contributing to spacemacs :purple_heart:, I have cherry picked your commit into develop.

**Pull request closed without merging**